### PR TITLE
Support for gjs 1.71.1 and gnome-shell 42

### PR DESCRIPTION
**Pull request by philippun1**
Makes dtp work with gnome 42 and keeps backwards compatibility with gnome 41.

### CODE REVIEW

File: utils.js [hunk 1]
```diff
@@ -80,19 +80,31 @@ var defineClass = function (classDef) {
...
-        '     callParent(...args) { ' +
-        '         let func = args.shift(); ' +
-        '         if (!(func === \'_init\' && needsSuper))' +
-        '             super[func](...args); ' +
-        '     }' +    
-        '})'
-    );
+    let C;
+    if (isGObject) {
+        C = eval(
```
[the remark this answers is not in the document]
Comment (miigotu) [reply utils.js:85]: Seems they just tried to avoid duplicates code or something. It's ugly for sure, and makes it look like an obfuscation.

**Comment by crosser** [comment]
I applied this as a patch to the source of the "official" package (`gnome-shell-extension-dash-to-panel_45-1`) and rebuilt it. Now it can be enabled. When it is enabled, ubuntu dash disappears, and the panel moves from the top to the bottom (as it should). But the _only_ element present in the panel is the "Show Applications" button. There are no active applications, no clock, no appindicators, the rest of the panel is blank.

File: utils.js [hunk 2]
```diff
@@ -347,7 +359,8 @@ var mergeObjects = function(main, bck) {
 var hookVfunc = function(proto, symbol, func) {
     if (Gi.hook_up_vfunc_symbol && func) {
         //gjs > 1.53.3
-        proto[Gi.hook_up_vfunc_symbol](symbol, func);
+        // todo
+        //proto[Gi.hook_up_vfunc_symbol](symbol, func);
```
Comment (philippun1) [inline utils.js:363]: Fixing this would probably resolve a lot of problems.

**Comment by philippun1** [comment]
Added patches from @crab2313, thanks!

**Comment by philippun1** [comment]
@charlesg99 I've tested this MR with gnome 41 (on Manjaro) and gnome 42 (on Ubuntu jammy), works on both.

**Comment by miigotu** [comment]
Removed ability to set custom corner size and style?

**Comment by hf29h8sh321** [comment]
There's an issue I'm experiencing. If I hover over something on the panel, the hover effect (preview, color, tooltip, etc) stays if I move my mouse up. The only way to get it to go away if to move mouse sideways to an empty part of the panel.

**Comment by Alexmitter** [comment]
So far the only issues I experience with the newest state of code are:
* panel does not detect if mouse was moved away from it.
* icon grid view is not shown

**Comment by philippun1** [comment]
@Alexmitter @hf29h8sh321 I cannot reproduce what you are describing. Could you please give some more details.

**Comment by philippun1** [comment]
AppMenu is now fixed and the prefs dialog is created properly. In my opinion this would be ready for merge.
Would be nice to have a version for the release of Ubuntu 22.04.
@charlesg99 What do you think?

**Comment by Alexmitter** [comment]
> @Alexmitter @hf29h8sh321 I cannot reproduce what you are describing. Could you please give some more details.

I will send you a video of this bug as soon as I am back home. As it is quite severe, I would ask to wait with merge until you have at least seen it.

**Comment by crosser** [comment]
> @Alexmitter @hf29h8sh321 I cannot reproduce what you are describing. Could you please give some more details.

FWIW, I cannot reproduce it either.

**Comment by Alexmitter** [comment]
> AppMenu is now fixed and the prefs dialog is created properly. In my opinion this would be ready for merge. Would be nice to have a version for the release of Ubuntu 22.04. @charlesg99 What do you think?

Idk what it fixed, but since pulling those two commits from the last hour I can not reproduce it anymore. Sorry to have bothered you.

**Comment by amovitz** [comment]
FWIW I've also tested these commits and haven't had any trouble with the latest commits.

**Comment by Alexmitter** [comment]
Bad news, I thought was fixed, but actually the bug just takes some time to appear. It should be easy to understand via the video  @philippun1 .

https://user-images.githubusercontent.com/6940479/159064371-9d3b7922-52de-47ee-b4f5-df460fb177ff.mp4


https://user-images.githubusercontent.com/6940479/159064832-5dab40c4-f2d3-4368-8c94-91edc00ad73d.mp4

**Comment by philippun1** [comment]
@Alexmitter I cannot reproduce this, works perfectly fine on my devices. Do you have any suspicious output in the log? (you can watch it via: journalctl /usr/bin/gnome-shell -f -o cat)

**Comment by Alexmitter** [comment]
@philippun1 
Here is the log: https://gist.github.com/Alexmitter/7068a8d8e1f8d497ac285eea1b4cc901
I do not know how to trigger it, it was fine after installing it and a while later it just started.

**Comment by crosser** [comment]
@philippun1 fyi built a jammy package from your branch (commit 133b6ef) and it works fine for me, including right-button menus and settings.

**Comment by philippun1** [comment]
@Alexmitter I think the interesting part is cut off at the beginning of the log. Did you find a way to reproduce it? Otherwise it will be very tricky to fix, I guess.
@crosser Great! Now we just need to get it into the jammy repo.

**Comment by Alexmitter** [comment]
@philippun1 I sadly really do not have a way to reproduce it, but it reliably happens after some time for me.

Fedora 36 Beta, 4 Screen Setup, idk what else could make a difference.

**Comment by hf29h8sh321** [comment]
> I cannot reproduce what you are describing. Could you please give some more details.

@philippun1 

This issue always occurs for me. I cannot determine the exact reason for the issue.

Here is some data:

nvidia 510.54

Xorg

gnome-shell 1:42.0-0.1

**Comment by hf29h8sh321** [comment]
Videos from @Alexmitter match the issue I experience.

**Comment by hf29h8sh321** [comment]
> Do you have any suspicious output in the log?

There are no new log outputs when the issue occurs.

**Comment by hf29h8sh321** [comment]
I tested again after restarting gnome-shell with all other extensions disabled. The same issue occurs.

**Comment by hf29h8sh321** [comment]
Issue also occurs on Wayland session.

**Comment by philippun1** [comment]
I've installed Fedora 36 in VirtualBox and could reproduce the app grid not showing. There is an API change in _getAppDisplayBoxForState with gnome 42.rc. Ubuntu uses 42.beta still, so it works properly (for now)... I pushed a fix for it.

I could not reproduce the hover effect though, I tried with 4 virtual monitors on wayland and xorg.

**Comment by Alexmitter** [comment]
@philippun1 I did a bit of extra try to get close to what triggers the bug and I found it, its firefox. I opened multiple applications and everything was fine until I opened firefox, then the bug instantly occurred.
Here is the log: https://gist.github.com/Alexmitter/5d1bf5089e5f219d7eb84b0812176a6b

**Comment by hf29h8sh321** [comment]
> I did a bit of extra try to get close to what triggers the bug and I found it, its firefox.

That's strange. The issue occurs for me no matter what is opened. It even occurs on startup with no applications started.

**Comment by philippun1** [comment]
If I open Firefox I can reproduce the issue. This is indeed very strange.

**Comment by hf29h8sh321** [comment]
With Firefox, does the issue occur when it's open, or does opening Firefox just start the issue? (Does the issue disappear after closing Firefox?)

**Comment by charlesg99** [comment]
> AppMenu is now fixed and the prefs dialog is created properly. In my opinion this would be ready for merge. Would be nice to have a version for the release of Ubuntu 22.04. @charlesg99 What do you think?

Thanks a lot for the update! I'll make sure to create a release before 22.04 hits.

**Comment by Alexmitter** [comment]
@philippun1 I have good (or bad news), I can reproduce the same bug without dash to panel.

**Comment by philippun1** [comment]
@Alexmitter Ok. Let's see how it behaves with the final gnome 42 version.

**Comment by hf29h8sh321** [comment]
@Alexmitter @philippun1 Has this bug been reported on gnome-shell? If not, I will create a bug report.

**Comment by hf29h8sh321** [comment]
I've received a reply on the bug report. I've tested a patch and it solves the issue.

File: overview.js [hunk 3]
```diff
@@ -590,6 +590,9 @@ var dtpOverview = Utils.defineClass({
                     workAreaBox.set_size(workArea.width, workArea.height);
     
                     params = [workAreaBox, searchHeight, dashHeight, workspaceAppGridBox]
+                    if (Config.PACKAGE_VERSION > '42.beta') {
```
Comment (crab2313) [inline overview.js:593]: This comparison doesn't work on 42.0

**Comment by crab2313** [comment]
Another issue. 

How to reproduce:

Keep clicking the 'Show Applications' button in the left corner. 

Log:

```
Mar 23 16:25:45 yoga14s gnome-shell[69042]: JS ERROR: Exception in callback for signal: showing: Error: No signal 'animation-done' on object 'Gjs_ui_iconGrid_IconGrid'
                                            onShownCb@/usr/share/gnome-shell/extensions/dash-to-panel@jderose9.github.com/taskbar.js:1251:44
                                            _onShowAppsButtonToggled/overviewShowingId<@/usr/share/gnome-shell/extensions/dash-to-panel@jderose9.github.com/taskbar.js:1248:25
                                            _emit@resource:///org/gnome/gjs/modules/core/_signals.js:114:47
                                            _animateVisible@resource:///org/gnome/shell/ui/overview.js:559:14
                                            show@resource:///org/gnome/shell/ui/overview.js:539:14
                                            _onShowAppsButtonToggled@/usr/share/gnome-shell/extensions/dash-to-panel@jderose9.github.com/taskbar.js:1265:31
```

Possible fix:

``` diff
diff --git a/taskbar.js b/taskbar.js
index d5d1f8b..838e0da 100644
--- a/taskbar.js
+++ b/taskbar.js
@@ -1240,15 +1240,6 @@ var taskbar = Utils.defineClass({
                 // runs if we are already inside the overview.
                 if (!Main.overview._shown) {
                     this.forcedOverview = true;
-                    let grid = AppDisplay._grid;
-                    let onShownCb;
-                    let overviewSignal = Config.PACKAGE_VERSION > '3.38.1' ? 'showing' : 'shown';
-                    let overviewShowingId = Main.overview.connect(overviewSignal, () => {
-                        Main.overview.disconnect(overviewShowingId);
-                        onShownCb();
-                    });
-
-                    onShownCb = () => grid.emit('animation-done');
                 }
 
                 //temporarily use as primary the monitor on which the showapps btn was clicked, this is
```

The spring animation is not used since gnome 40 and removed in gnome 42.

**Comment by philippun1** [comment]
@crab2313 I removed the spring animation code. There is some more code for the spring animation, which could be removed as well, I will take a look at it.

**Comment by Nilvus** [comment]
Hi, thanks for working on Gnome 42 support. I use Debian Sid and even with your MR, it doesn't work.

See error on Gnome Extensions:

image

Note that version of this MR is set to 40 (on Gnome extensions, actual one is 45 so Gnome extensions want to update it).

image

**Comment by philippun1** [comment]
@Nilvus Debian SID already has gjs 72 but still gnome 41. It will work once it has moved to gnome 42.

**Comment by rolandixor** [comment]
How do I install this on Ubuntu 22.04? I tried building from the main branch, but it didn't work.

**Comment by miigotu** [comment]
> @Nilvus Debian SID already has gjs 72 but still gnome 41. It will work once it has moved to gnome 42.

That doesn't sound like this PR is ready. The conditionals should have allowed dash-to-panel to - continue working on gnome-shell 40+, regardless of gjs version. Backwards compatibility is important.

**Comment by Nilvus** [comment]
Thanks for your reply @philippun1. Though, I agree with @miigotu.

**Comment by philippun1** [comment]
@Nilvus I updated the condition check. Please try it again. I only tested it on Ubuntu, Fedora, Manjaro and Gnome OS...

**Comment by Nilvus** [comment]
> @Nilvus I updated the condition check. Please try it again. I only tested it on Ubuntu, Fedora, Manjaro and Gnome OS...

All good. Thanks for the quick fix. It works. Remains just an annoying thing:

Version is set at 40 with that MR and so Gnome Extensions want to update extensions for version actually on extensions.gnome.org (version 45). So, when restarting Gnome Shell, dash to panel is updated and so doesn't work again. I update it manually on extension install folder as metadata.json has 9999 version with MR in source code, so I don't know how 40 number is set. Hope I'm clear here.

**Comment by philippun1** [comment]
@Nilvus I just installed it with "make install" and I have the correct version.

**Comment by nikodunk** [comment]
I've been daily driving a clone of your branch on Fedora 36 Beta side-mounted for 2 days now (make install) and it works well for me! Thanks so much for the hard work on this.

**Comment by Nilvus** [comment]
> @Nilvus I just installed it with "make install" and I have the correct version.

Strange, i do exactly the same after doing a git clone of your repository then git checkout to your branch and each time I have version 40... And I always start with also no dash-to-panel folder in extensions folder on my personal folder.

File: utils.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (miigotu) [reply utils.js:85]: Is there a way to make this native without eval? It makes automated code analysis difficult and unreliable.

**Review by miigotu — COMMENTED** [review]
Honestly, I don't see any code that should break backwards compatibility. Maybe checking against gjs version is needed somewhere as well?

**Comment by philippun1** [comment]
@Nilvus I forgot to push the tags from the official repo to my repo, _make install_ should work properly now.

I couldn't reproduce it, because I have both remotes added and always pull the master branch from the official repo.

**Comment by Nilvus** [comment]
> @Nilvus I forgot to push the tags from the official repo to my repo, _make install_ should work properly now.

I confirm that it is now ok. On git pull I have seen new tag to v41 from v45 appear! Thanks for quick fix here. This seems now ready to release.

**Comment by palWorx** [comment]
Maybe something is wrong with intellihide, because after git clone and git pull gjs i have v45 and the extension isn't crashing anymore but there is a panel for some milliseconds and then it's gone.
When i set option to put the panel on the left and use 100% width, i see some dots from the app grid on the left lower corner, means the panel is way off outside my screen. It does not matter if autohide is on or off.

**Comment by nexryai** [comment]
In my environment, intellihide does not work at all and enabling it does not change anything...

**Comment by palWorx** [comment]
@philippun1 
Excuse me for not giving up in hope of a working extension.

The moment i activate the extension journal shows:
Can't update stage views actor <unnamed>[<MetaWindowGroup>:0x55f011208350] is on because it needs an allocation.
Can't update stage views actor <unnamed>[<MetaWindowActorX11>:0x55f013766750] is on because it needs an allocation.
Can't update stage views actor <unnamed>[<MetaSurfaceActorX11>:0x55f0137686c0] is on because it needs an allocation.
This message repeats over and over again.

I also found several times: Object .Gjs_dash-to-panel_jderose9_github_com_utils_DashToPanel_TaskbarAppIcon (0x55f0124179c0), has been already disposed — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.

Debian sid:
gnome-core                                                 1:42+1
gnome-session                                            42.0-1
gnome-session-bin                                      42.0-1
gnome-session-common                             42.0-1
gnome-settings-daemon                              42.1-1
gnome-settings-daemon-common               42.1-1
gnome-shell                                                 42.0-2
gnome-shell-common                                  42.0-2
gnome-shell-extension-appindicator            42-1
gnome-shell-extension-prefs                       42.0-2

**Comment by philippun1** [comment]
@palWorx Yes, I have these as well. But in my opinion this is not a blocker for this pull request. I will try to fix it soon.

**Comment by miigotu** [comment]
> @palWorx Yes, I have these as well. But in my opinion this is not a blocker for this pull request. I will try to fix it soon.

@philippun1 How can I help you fix those issues? A more complete PR might be in order this time, to catch up with the maintenance. Usually I am in favor of small, focused pulls, but maybe we can get it fully functional and cleaned up in one go?

Should I create an organization for the few of us that want to continue on with making this project great or should I just add you to write permissions on my fork? (If you are interested in working with me on a fork)

**Comment by palWorx** [comment]
@philippun1 
Found my problem. Its the extension "Just Perfection", damn thats bad news. I deactivated all extenions and only this one made the panel vanish. It doens't matter what option i enable or disable in "Just Perfection". Well it seems it's just not that perfect in the end.

Perhaps, i kindly ask, only if you have time, i would be very happy, if you could have a look into it and why the two won't coexist in Gnome 42 and gjs 1.72.0.

**Comment by palWorx** [comment]
@philippun1 
Ok, last time bothering you. I tidied up a bit and removed in dconf (reset recursive) "Dash-to-dock", ja sorry in between i betrayed you, made a F2+r and after that i pressed standard in "Just Perfection" (which did not work before) and now both are working together.
So thanks for a great app, i am happy again.

**Comment by charlesg99** [comment]
Thanks Philippun1 and @crab2313!

**Pull request merged**